packages/hub/tasks/process-reward-root.ts

```diff
@@ -0,0 +1,186 @@
+    try {
+      if (rows.length > 0) {
+        const proofsQuery = `
+      INSERT INTO reward_proofs(
+        reward_program_id, payee, leaf, payment_cycle, proof, token_type,  valid_from, valid_to, explanation_id, explanation_data
+      ) VALUES %s;
+    `;
+        const proofsSql = pgFormat(proofsQuery, rows.join());
+
+        await this.databaseManager.performTransaction(db, async () => {
+          await db.query(indexSql);
+          const r = await db.query(proofsSql);
+          log.info(`Insert ${r.rowCount} notification_types rows`);
+        });
+      } else {
+        await db.query(indexSql);
+        //will rememeber root  as indexed
+        //although no proofs (cause their expired) are stored in db
+      }
+    } catch (e: any) {
+      if (e.code == '23505') {
+        // do not scream errors when indexing
+        log.info(`rewardProgramId: ${file.rewardProgramId}, paymentCycle: ${file.paymentCycle} is already indexed`);
+        return;
+      }
+    }
+  }
+}
+
+declare module '@cardstack/hub/tasks' {
+  interface KnownTasks {
+    'process-reward-root': ProcessRewardRoot;
+  }
+}
+
+const FIELDS_EXCEPT_EXPLANATION_DATA =
+  'rewardProgramID,paymentCycle,validFrom,validTo,tokenType,payee,root,leaf,proof,explanationId';
+
+const queryParquet = async (
+  s3Client: S3Client,
+  file: S3FileInfo,
+  expression?: SelectObjectContentCommandInput['Expression']
+): Promise<Proof[]> => {
+  let records: Proof[] = [];
+  const params: SelectObjectContentCommandInput = {
+    Bucket: 'cardpay-staging-reward-programs',
+    Key: `rewardProgramID=${file.rewardProgramId}/paymentCycle=${file.paymentCycle}/results.parquet`,
+    ExpressionType: 'SQL',
+    Expression: expression ?? 'SELECT * FROM S3Object',
+    InputSerialization: {
+      Parquet: {}, //no parameters needed here
+    },
+    OutputSerialization: {
+      JSON: {
+        RecordDelimiter: '\n', //each row can be separated by splitting \n
+      },
+    },
+  };
+  try {
+    const command = new SelectObjectContentCommand(params);
+    const s3Data = await s3Client.send(command);
+
+    if (!s3Data.Payload) {
+      throw new Error('payload undefined');
+    }
+    const events: AsyncIterable<SelectObjectContentEventStream> = s3Data.Payload;
+    for await (const event of events) {
+      if (event?.Records) {
+        if (event?.Records?.Payload) {
+          const record = Buffer.from(event.Records.Payload).toString('utf8');
+          const os = record.split('\n');
+          os.map((o) => {
+            if (o != '') {
+              const b: any = JSON.parse(o);
+              if (!('explanationData' in b)) {
+                b.explanationData = '{}';
+              }
+              // when proof array serialize to json, it becomes [{item:... }. {item: ...}]
+              // need to transform to array of hex strings
+              b.proof = b.proof.length > 0 ? b.proof.map((o: any) => '0x' + o.item) : ['0x'];
```

Comment:
I think the empty case should be [], otherwise the client will get a proof array with a single item in.

Reply:
This has been the pattern we've been using actually. When there is a single reward, proof = ["0x"] on returned by the api. When there is two rewards proof= ["0xbla"]. The sdk handles it as an array in both instances. 

I renamed proof to proofBytes for clarity (and also bcos of prisma related issue on pluralization)

Reply:
oh ok, thanks!